tarexport: optimize logging in Load function

tdakkota opened the pull request:
**- What I did**

I reduced allocations related to the logging.

**- How I did it**

This PR replaces `fmt.Sprintf` call with `fmt.Fprintf`.

The `w.Write([]byte(fmt.Sprintf(...)))` pattern is quite inefficient, it allocates
3x more than fmt.Fprintf variant.

Also, it replaces `string` with a `bytes.Buffer` to reduce `string` -> `[]byte` conversion
overhead.

**- How to verify it**

This PR does not contain any functional changes. I think special verification is not neccessary.

**- Description for the changelog**
tarexport: optimize logging in Load function

**- A picture of a cute animal (not mandatory but encouraged)**

<details><summary> Cute cat </summary>
<img src="https://memepedia.ru/wp-content/uploads/2020/10/big-floppa-meme.png"></img>
</details>

thaJeztah commented on image/tarexport/load.go:
```diff
@@ -129,7 +134,7 @@ func (l *tarexporter) Load(inTar io.ReadCloser, outStream io.Writer, quiet bool)
 			}
 			l.setLoadedTag(ref, imgID.Digest(), outStream)
-			outStream.Write([]byte(fmt.Sprintf("Loaded image: %s\n", reference.FamiliarString(ref))))
+			fmt.Fprintf(outStream, "Loaded image: %s\n", reference.FamiliarString(ref))
```
I guess this could also just concatenate (if we want to get rid of `sprintf` / `fprintf`.

```go
outStream.Write([]byte("Loaded image ID: " + reference.FamiliarString(ref) + "\n"))
```

Looks like that's slightly faster

```
BenchmarkWithFprintf2-8   	27867684	        44.11 ns/op	      83 B/op	       0 allocs/op
BenchmarkWithConcat2-8    	29602640	        37.75 ns/op	      49 B/op	       0 allocs/op
```

thaJeztah reviewed (changes requested):
left some comments inline

Slightly wondering if this is premature optimization, or if you were running into actual performance issues related to this 😅

tdakkota committed:
tarexport: optimize logging in Load function

This commit replaces fmt.Sprintf call with io.WriteString
and string concatenation.

Also, it replaces string with a bytes.Buffer to reduce string -> []byte conversion
overhead.

Signed-off-by: tdakkota <tanc13@yandex.ru>

tdakkota commented on image/tarexport/load.go:
```diff
@@ -116,7 +119,9 @@ func (l *tarexporter) Load(inTar io.ReadCloser, outStream io.Writer, quiet bool)
 		}
-		imageIDsStr += fmt.Sprintf("Loaded image ID: %s\n", imgID)
+		imageIDsStr.WriteString("Loaded image ID: ")
+		imageIDsStr.WriteString(imgID)
```
> Looks like this needs to be converted to a string here;
> 
> ```go
> 		imageIDsStr.WriteString(imgID.String())
> ```



Sure, don't know how I missed that.


> 
> That said, I guess this could also just concatenate the strings (to avoid the `Sprintf`);
> 
> ```go
> imageIDsStr.WriteString("Loaded image ID: " + imgID.String() + "\n")
> ```
> 
> A quick bench between using the code in this PR, and `imageIDsStr.WriteString("Loaded image ID: " + imgID.String() + "\n")`;
> 
> ```
> BenchmarkWithBuffer-8   	25752823	        58.00 ns/op	      89 B/op	       0 allocs/op
> BenchmarkWithConcat-8   	31360424	        37.99 ns/op	      46 B/op	       0 allocs/op
> ```

It seems Go optimizes string concatination. Maybe it allocates a buffer of exact size unlike `bytes.Buffer`.

thaJeztah reviewed (approved):
LGTM

fredericdalleau commented on image/tarexport/load.go:
```diff
@@ -116,7 +119,7 @@ func (l *tarexporter) Load(inTar io.ReadCloser, outStream io.Writer, quiet bool)
 			return err
 		}
-		imageIDsStr += fmt.Sprintf("Loaded image ID: %s\n", imgID)
+		imageIDsStr.WriteString("Loaded image ID: " + imgID.String() + "\n")
```
Hi, you could completely avoid concatenation by using :
```
imageIDsStr.WriteString("Loaded image ID: ")
imageIDsStr.WriteString(imgID.String())
imageIDsStr.WriteString("\n")
```
WDYT?
